### ISSUE TITLE
Axioms of choice

### ISSUE THREAD

**Issue by DanGrayson** [UniMath/UniMath#568]
I'd like to formulate two axioms of choice, one which says that a surjection between sets
has a section, and another that says any type is the target of a surjective map from a set.
The logical place to put these, since they relate to sets, is at the end of the file Sets.v.  In
Propositions.v we formulate LEM, not as a Coq Axiom, but simply as a type, so that any statements using it have to state it explicitly as a hypothesis.  I'd like to do the same 
with the two axioms of choice.  Is there any discussion?  (Or should they just be axioms? 
(That would hide the dependency of theorems on the axioms, which might not be good.)

(The second axiom of choice is consistent with univalence, because given a simplicial
set, its set of vertices is a set, and the inclusion map is surjective up to homotopy.)

I could also put various lemmas about alternative formulations there, or additional
axioms of choice, if desired.  A proof of Zorn's lemma would be good, too, but it
would go in the file OrderedSets.v.

A future use for the second axiom of choice above comes from higher algebraic K-theory,
which I will eventually formalize.  What is often done is to take a category arising in algebra,
make a new category from it, and then take the nerve of that category (as a simplicial set).  For the
latter step, we need the objects of the category to form a set.  With the second axiom of choice,
we can, for each category, replace it by an equivalent one whose objects form a set.

The concrete examples of exact categories arising in algebraic K-theory all come in both flavors:
a precategory with a set of objects, and a univalent category.  For example, consider the
category of finitely presented abelian groups versus the category of finitely presentable
groups.  The objects in the former category come with a finite presentation as part of the data
and the objects in the latter come just with a proof that such a presentation exists. This  observation shows that the second axiom of choice is not really 
needed for doing algebraic K-theory.

**Comment by DanGrayson** [comment]
PS: Are there any lemmas about homotopy pullbacks in Foundations?

**Comment by vladimirias** [comment]
Yes, in Basics/PartA.v and maybe elsewhere (you should be able to use grep after you get the names of the first definitions from PartA).

I’ll comment on your suggestions about axioms of choice tomorrow.

**Comment by DanGrayson** [comment]
I looked in PartA and couldn't find "pull" or "pullback".

**Comment by vladimirias** [comment]
It is called homotopy fiber squares.

While you are at it, could you please add a few comment likes there saying that homotopy a pullback is a synonym?

**Comment by vladimirias** [comment]
Sorry, I did not wake up yet. “… could you please add a few comments there, like saying that “a homotopy pullback" is a synonym?” is the intended reading of my request.

**Comment by DanGrayson** [comment]
Okay, will do.  I found it here:
```
(** *** Homotopy fiber products *)

Definition hfp {X X' Y : UU} (f : X -> Y) (f' : X' -> Y)
  := total2 (fun xx' : dirprod X X' => (f' (pr2 xx')) = (f (pr1 xx'))).
```
The equation looks backwards, by the way.  Perhaps I should fix that, too.

**Comment by vladimirias** [comment]
I open a separate issue for it.

**Comment by DanGrayson** [comment]
PS: What would be a good name for the axiom of choice that says we can find a surjective
map from a set to each type?

**Comment by vladimirias** [comment]
This follows from the axiom of choice that says that given a function from a type to a set that is surjective it has a section (must be in the truncated form).

One applies it to the projection T -> pi_0(T).

**Comment by DanGrayson** [comment]
Yes -- that reduces the number of axioms from 1 to 2.

**Comment by DanGrayson** [comment]
Vladimir mentioned to me today that he's unsure whether we can prove Zorn's lemma and Zermelo's well ordering theorem from the axiom of choice in UniMath, but I'm confident we can.  So I'll try that next.

**Comment by DanGrayson** [comment]
This will be resolved by pr #637

**Comment by DanGrayson** [comment]
Closed by #637
